API: Created user profile GET and PATCH endpoints

andrii-zhygalko opened the pull request.

andrii-zhygalko committed:
API: Created user profile GET and PATCH endpoints

andrii-zhygalko commented on src/db/models/User.js:
Цей файл має бути задіяний і для auth ендпоінтів, тому висока вірогідність конфлікту, коли прийде PR з цими ендпоінтами. Якщо так, потрібно буде вирішити який варіант залишити.

maksymPr424 replied:
Да, це ще треба буде перевірити

maksymPr424 replied:
Я не впевнений, що треба робити дефолтну картинку так. Тому що вона змінюється залежно від теми. Краще буде зберігати варіанти на фронті, а з беку відправляти тільки якщо юзер щось додав, або підтяглясь із OAuth. Пілправ цей момент

andrii-zhygalko commented on src/validation/user-schemas.js:
Цей файл теж може пересікатись з auth. Якщо прилетить PR з таким же ім'ям - треба буде вирішити яку версію лишити, якщо з іншим - то уніфікувати, щоб не було дублікатів

andrii-zhygalko commented on src/controllers/user.js:
```diff
@@ -0,0 +1,68 @@
+import { env } from '../utils/env.js';
+import hashPassword from '../utils/hashPassword.js';
+const cloudinaryFolder = env('CLOUDINARY_FOLDER');
+export const getUserByIdController = async (req, res) => {
+  // const userId = req.user._id;
+  const userId = req.params.userId;
+  const user = await userServises.findUser({
+    _id: userId,
+  });
+
+  if (!user) {
+    throw createHttpError(404, 'User not found');
+  }
+
+  const { photoUrl, theme, name, email } = user;
+
+  res.json({
+    status: 200,
+    message: 'Successfully retrieved user profile',
+    data: { name, email, theme, photoUrl },
+  });
+};
```
На рахунок цього контроллера я не впевнений, чи він потрібний. Коли юзер клікає на свій профіль, то відкривається модалка, і я так думав, що робиться запит, який прийде на цей контроллер і фронтенд отримає поточні дані про користувача для цієї модалки. А потім думаю, по суті ж поточні дані і так будуть в сторі редаксу, тоді цей GET буде зайвий. Він-то нічого не зламає, просто може бути зайвий код

maksymPr424 replied:
Я думаю що саме котроллер не потрібний. Тому що в ідеалі ми булемо надсилати данні про користувача разом з логіном/рефрешом. Сервіс по люьому знадобиться для використання в інших контроллерах
```suggestion
export const getUserByIdController = async (req, res) => {
  // const userId = req.user._id;
  const userId = req.params.userId;
  const user = await userServises.findUser({
    _id: userId,
  });

  if (!user) {
    throw createHttpError(404, 'User not found');
  }

  const { photoUrl, theme, name, email } = user;

  res.json({
    status: 200,
    message: 'Successfully retrieved user profile',
    data: { name, email, theme, photoUrl },
  });
};
```

andrii-zhygalko commented:
Привіт, глянь коменти вище, щодо можливих конфліктів. Також я створив трохи нових файлів 
- для функціоналу завантаження аватарки на cloudinary
- враппер для контроллерів (потенційний конфлікт, якщо ще хтось його створив)
- middleware для завантаження фото юзера в тимчасову папку, утиліта хешування пароля (при оновленні профілю користувача). 
- файли з константами

maksymPr424 commented on src/validation/user-schemas.js:
Пододавай в емейли також валідацію на кількість символів. Впиши якось від 4 до умовно 20-30. Щоб нам гавна не накидали туда. А так все супер

maksymPr424 commented on src/constants/users-constants.js:
Видали цей файл, він пустий

maksymPr424 reviewed (changes requested):
Я відмітив там пару моментів, але вони не значні. Загалом все круто. Я ек знаю, які роути ти можеш ще зробити. Хіба що доку для апдейта

andrii-zhygalko committed:
Merge remote-tracking branch 'origin/main' into feature/user-profile-endpoints

andrii-zhygalko committed:
updated env example, email validation, and PATCH endpoint docs

- Added `.env.example` for required environment variables (for cloudinary)
- Enforced email min/max length validation
- Documented PATCH `/user/:id` endpoint in Swagger

andrii-zhygalko commented:
Зробив новий push (назва коміту `updated env example, email validation, and PATCH endpoint docs`
- видалив зайвий контролер
- оновив swagger для ендпоінта user/:id для методу PATCH (ендпоінд документації `/api-docs`)
- оновив валідацію email по кількості символів (min/max)

В контролері змінна `userId `зараз береться з `req.params.userId`, тому що `req.user._id` поки що в req не додається, бо middleware `authenticate` ще не готова, так як має залежності з іншими файлами в розробці.

maksymPr424 commented on swagger/paths/user/{id}/post.yaml:
Тут обговорили, що помилка. Я все інше перевірив, роут працює, все добре

maksymPr424 reviewed (changes requested):
Фікси доку і приймаю)

andrii-zhygalko committed:
minor changes in swagger docs and variables naming

andrii-zhygalko commented on docs/openapi.yaml:
Виправив swagger

maksymPr424 replied:
Поміняй тип на aplication/json і я прийму

maksymPr424 reviewed (changes requested):
Зміни тип в свагері на aplication/json і я прийму
